### PR TITLE
Allow using a custom service account on an ipc-registration.

### PR DESCRIPTION
**Pull request by ensonic**
One can already configure a custom service account by using the annotation on the configmap. The api will be added in teh next changes.

### CODE REVIEW

**Comment by ensonic** [comment]
Mostly FYI (and to save my changes).

File: src/go/cmd/token-vendor/app/tokenvendor.go [hunk 1]
```diff
@@ -36,7 +36,7 @@ type PubKeyRepository interface {
 	// LookupKey retrieves the public key of a device from the repository.
 	// An empty string return indicates that no key exists for the given identifier or
 	// that the device is blocked.
```
Comment (csieber) [inline src/go/cmd/token-vendor/app/tokenvendor.go:38]: Can we document the two return values here?

File: src/go/cmd/token-vendor/repository/k8s/k8s.go [hunk 2]
```diff
@@ -63,21 +65,22 @@ func (k *K8sRepository) ListAllDeviceIDs(ctx context.Context) ([]string, error)
 // The public key is stored under a specific key in the configmap. If the configmap
 // does not exist, we return an empty string. For any other error or a malformed
 // configmap we return an error.
-func (k *K8sRepository) LookupKey(ctx context.Context, deviceID string) (string, error) {
+func (k *K8sRepository) LookupKey(ctx context.Context, deviceID string) (string, string, error) {
```
Comment (csieber) [inline src/go/cmd/token-vendor/repository/k8s/k8s.go:68]: I think we should rename this to LookupConfig and return a config struct here. However that can be a follow-up.

File: src/go/cmd/token-vendor/tokensource/gcp.go [hunk 3]
```diff
@@ -44,15 +44,19 @@ func NewGCPTokenSource(ctx context.Context, client *http.Client, project, sa str
 // Token returns an access token for the configured service account and scopes.
 //
 // API: https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateAccessToken
-func (g *GCPTokenSource) Token(ctx context.Context) (*TokenResponse, error) {
+func (g *GCPTokenSource) Token(ctx context.Context, saName string) (*TokenResponse, error) {
 	req := iam.GenerateAccessTokenRequest{Scope: g.scopes}
+	resource := g.resource
```
Comment (csieber) [inline src/go/cmd/token-vendor/tokensource/gcp.go:49]: can we rename resource to defaultResource or similar?

Comment (csieber) [reply src/go/cmd/token-vendor/tokensource/gcp.go:49]: And match the name with the parameter name

**Review by csieber — APPROVED** [review]

File: src/go/cmd/token-vendor/app/tokenvendor.go (hunk 1 above)
Comment (ensonic) [reply src/go/cmd/token-vendor/app/tokenvendor.go:38]: I only uploaded the PR to save it. What if I introduce a struct Key that has all the values?

File: src/go/cmd/token-vendor/tokensource/gcp.go (hunk 3 above)
Comment (ensonic) [reply src/go/cmd/token-vendor/tokensource/gcp.go]: Done.

File: src/go/cmd/token-vendor/repository/k8s/k8s.go (hunk 2 above)
Comment (ensonic) [reply src/go/cmd/token-vendor/repository/k8s/k8s.go]: Done

File: src/go/cmd/token-vendor/app/tokenvendor.go (hunk 1 above)
Comment (ensonic) [reply src/go/cmd/token-vendor/app/tokenvendor.go]: Done

**Comment by ensonic** [comment]
PTAL. Maybe as a followup I pass the defaultSA to the tokenvendor app and not the token source. Then the token source is just receiving the sa-name and getting a token for it. That also fixes the logging issue. WDYT?

File: src/go/cmd/token-vendor/repository/repository.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+package repository
+
+import (
+	"context"
+)
+
+// Key holds data + metadata of a public key entry
+type Key struct {
+	PublicKey string
+	SAName    string
```
Comment (csieber) [inline src/go/cmd/token-vendor/repository/repository.go:25]: maybe document this as optional

**Review by csieber — APPROVED** [review]

File: src/go/cmd/token-vendor/app/tokenvendor.go [hunk 5]
```diff
@@ -95,48 +96,48 @@ func (tv *TokenVendor) GetOAuth2Token(ctx context.Context, jwtk string) (*tokens
...
 	if err != nil {
 		return nil, err
 	}
-	cloudToken, err := tv.ts.Token(ctx)
+	cloudToken, err := tv.ts.Token(ctx, sa)
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to retrieve a cloud token for device %q", deviceID)
 	}
-	slog.Info("Handing out cloud token", slog.String("DeviceID", deviceID))
+	slog.Info("Handing out cloud token", slog.String("DeviceID", deviceID), slog.String("ServiceAccount", sa))
```
[the remark this answers is not in the document]
Comment (csieber) [reply src/go/cmd/token-vendor/app/tokenvendor.go:135]: The other option could be a function like `func (context.Context, string) (*TokenSource, error)`

**Comment by csieber** [comment]
> PTAL. Maybe as a followup I pass the defaultSA to the tokenvendor app and not to the token source. Then the token source is just receiving the sa-name and getting a token for it. That also fixes the logging issue. WDYT?

Maybe two arguments. `func(contect.Context, string) (*TokenSource, error)` and default service account.

**Pull request merged**